proto/palomachain/paloma/consensus/tx.proto

```diff
@@ -20,12 +21,14 @@ service Msg {
 }
 
 message MsgAddMessagesSignatures {
+  option (cosmos.msg.v1.signer) = "metadata";
```

Comment:
Is this needed? If so, why - and where can I see the implication of this change in this PR?

Doesn't seem to affect the generated types. According to the docs:
```
option (cosmos.msg.v1.signer) identifies X as the field that will serve as the signer. At compilation time, the SDK will automagically pick up the value of this annotation to have your message implement sdk.Msg.GetSigners.
```

Then again, `metadata` is a composed struct, not an address. Does this change make sense and if so, where can I see it?

Reply:
Txns were failing with error related to signer, so we added this to define signer.

Reply:
Can you give the exact error you ran in to? Was it simply enforcing you to specify this information? Because the added options here have not resulting in a change of generated code. So I wonder where this error would have come from and why this seems to fix it.

Reply:
[This](https://github.com/cosmos/cosmos-sdk/blob/9a4091478c5e3c21e99670de6eda596ad7d96b29/x/tx/signing/context.go#L160) was the error we got while executing transactions. Since the custom msg type [here](https://github.com/palomachain/paloma/pull/1069/files/bcdf4d248c51caa2898f58d4b42e0d91add59395#diff-e70526f754289da7f9ba89f1c44add9c03d5bb6e3fedcfa7e6b0ec74cfdc7d1bR21) is `metadata`

Reply:
When did you encounter the error? During code generation? During compilation? Runtime?

Also, are you saying you encountered this error: `a custom signer function as been defined for message %s which already has a signer field defined with (cosmos.msg.v1.signer)` BEFORE you added the `cosmos.msg.v1.signer` option to our message definitions? That doesn't seem to make sense, as the error will only be thrown if this option is present & a custom signer function exists (which it should).

The reason why I'm pressing so hard on this is because our antehandler which handles our custom solution for using multiple signature keys relies heavily on this `GetSigners()` signature, and we need to ensure that there is no difference in implementation.

Reply:
While executing the transaction we got this error.

Reply:
Ah, okay. So this is even worse, as it injects the `GetSigner()` function dynamically using reflection...

Which message(s) was/were in that transaction? 

Looking at the implementation [here](https://github.com/cosmos/cosmos-sdk/blob/9a4091478c5e3c21e99670de6eda596ad7d96b29/x/tx/signing/context.go#L159), what happens is:

```
IF cosmos.msg.v1.signer is part of message AND custom getter is defined => Throw error
```

So as long as `cosmos.msg.v1.signer` isn't defined on the messages, this error should not even occur. We don't really want to use dynamic injection of this method, as we need rely on our custom getters.

Reply:
Almost all transactions threw the same error, especially  `palomad tx valset add-external-chain-info-for-validator`. Btw the change [here](https://github.com/palomachain/paloma/blob/bcdf4d248c51caa2898f58d4b42e0d91add59395/x/paloma/ante.go#L96) has impact on test cases so I commented it. Does that has any impact on above scenario as well?

Reply:
Yes, we need the `libmeta` implementation to work, otherwise we'll run into problems with Pigeons using multiple signing keys.

Reply:
Well, we tried out this but ran into test case issues on using `libmeta`. It would help you can look at this `libmeta` usage and the ante_test.go to pass with that usage.

Reply:
I can take a look. What does your test setup look like for this? Just `make test`?

Reply:
`make test` would compile all module tests, so specifically to address this `libmeta` issue you might need to run `go test` inside `x/paloma` folder

Reply:
@deepan95dev I made the necessary change and all tests are green. I'm not sure where this error would come in that you spoke of? I pushed the changes to the branch, please can you try and reproduce this error?

Reply:
sure, I will try to attach the error screenshot in txn, to address the issue better

Reply:
Here is the error message if we remove the "metadata" signer   @byte-bandit  @atheeshp 
![TXN_ERROR](https://github.com/palomachain/paloma/assets/150890063/97fc88b7-d35f-493e-ba85-6e222a2a20a8)

Reply:
Okay, `no cosmos.msg.v1.siger option found`. That's a different error than you linked me to here: https://github.com/palomachain/paloma/pull/1069#discussion_r1462733799

It makes sense. The code is trying to figure out which property represents the signing authority for the message, now that the official `GetSigners()` method was removed.

What happens when you use `DefineCustomGetSigners()` as hinted at by the error log?

---

Just FYI, I also spent a few hours today running the latest version of this branch and the changed Pigeon on a private test net, and there seem to be some changes with regards to setup and configuration as well. 

The last error I ran into was `invalid coin denomination: got ugrain, expected stake`, at which I stopped for now. Just an FYI that there's more issues ahead after resolving this.

Reply:
Yeah, I should have pointed [this error message](https://github.com/cosmos/cosmos-sdk/blob/d8fb76d4c2671d1bd8ab656a66752091589bb4ad/x/tx/signing/context.go#L119) in above [#1069 (comment)](https://github.com/palomachain/paloma/pull/1069#discussion_r1462733799).

So this error message was fixed with `customGetSigners` [here](https://github.com/palomachain/paloma/blob/481fc23e60ec72416b23381556849894ff7975b2/app/app.go#L308)  which requires `cosmos.msg.v1.siger option` in proto files, so we need to revert [this commit](https://github.com/palomachain/paloma/pull/1069/commits/481fc23e60ec72416b23381556849894ff7975b2). 


Also can you list down the steps to test along with Pigeon App?

Reply:
> The last error I ran into was `invalid coin denomination: got ugrain, expected stake`, at which I stopped for now. Just an FYI that there's more issues ahead after resolving this.

For this issue we used `--default-denom ugrain` flag while initiating genesis.

Reply:
Please read up on these changes, this is an important change for us and it needs to be well understood. The `cosmos.msg.v1.signer` option is the default CosmosSDK implementation on how to extract signing information from a message. It is needed because `GetSigners()` is no longer part of the `Msg` interface. 

The background of this is easy. With the new option, retrieving the signers is automated. You no longer have to define it yourself for every message, which in theory is great.

However, Paloma uses multiple fields for signature information and aggregates those into one flattened list of signers. The arithmetic logic for this is in `libmeta`. That's why it is **not** enough to simply read one property of the message.

The default `cosmos.msg.v1.signer` simply extracts tells the framework which field to extract when looking at signing infos. This will **not** work with our implementation, because - again - Paloma needs to inspect both the `Creator` **and** the `Signers` field on the metadata.

Luckily, we're not the only ones with this problem, i.e. see [here](https://github.com/cosmos/cosmos-sdk/issues/16112). So Cosmos has introduced this `customGetSigners` concept which allows you to specify your own functions for signer identification, and override the `cosmos.msg.v1.signer` implementation.

Specifying `cosmos.msg.v1.signer` with `customGetSigners` as you have done is taking the whole idea of it Ad Absurdum. Instead, you want to feed in your own retriever function, which again makes use of `libmeta`.

Please also do your own research into this.

---

As for the testing alongside Pigeon, we have a full run book for you, but I think it makes sense to fix all of these issues before we kick off E2E testing.